nearcore/src/cold_storage.rs

```diff
@@ -0,0 +1,183 @@
+use std::sync::{atomic::AtomicBool, Arc};
+
+use near_chain::types::Tip;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::{hash::CryptoHash, types::BlockHeight};
+use near_store::{
+    cold_storage::{update_cold_db, update_cold_head},
+    db::ColdDB,
+    DBCol, NodeStorage, Store, FINAL_HEAD_KEY, HEAD_KEY,
+};
+
+use crate::{NearConfig, NightshadeRuntime};
+
+/// A handle that keeps the state of the cold store loop and can be used to stop it.
+pub struct ColdStoreLoopHandle {
+    join_handle: std::thread::JoinHandle<()>,
+    keep_going: Arc<AtomicBool>,
+}
+
+impl ColdStoreLoopHandle {
+    pub fn stop(self) {
+        self.keep_going.store(false, std::sync::atomic::Ordering::SeqCst);
+        match self.join_handle.join() {
+            Ok(_) => {
+                tracing::debug!(target:"cold_store", "Joined the cold store loop thread");
+            }
+            Err(_) => {
+                tracing::error!(target:"cold_store", "Failed to join the cold store loop thread");
+            }
+        }
+    }
+}
+
+/// The status of the cold store loop indicates if the loop should sleep before
+/// the next copy attempt or if it should actively continue copying.
+enum ColdStoreLoopStatus {
+    Sleep,
+    Continue,
+}
+
+// Checks if cold store head is behind the final head and if so copies data
+// at height equal to the cold store head height + 1.
+fn cold_store_copy(
+    hot_store: &Store,
+    cold_store: &Store,
+    cold_db: &Arc<ColdDB>,
+    genesis_height: BlockHeight,
+    runtime: &Arc<NightshadeRuntime>,
+) -> anyhow::Result<ColdStoreLoopStatus> {
+    let cold_head = cold_store.get_ser::<Tip>(DBCol::BlockMisc, HEAD_KEY)?;
+    let cold_head_height = match cold_head {
+        Some(cold_head) => cold_head.height,
+        None => genesis_height,
+    };
+
+    // If FINAL_HEAD is not set for hot storage though, we default it to 0.
+    let hot_final_head = hot_store.get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?;
+    let hot_final_head_height = match hot_final_head {
+        Some(hot_final_head) => hot_final_head.height,
+        None => genesis_height,
+    };
+
+    tracing::debug!(target: "cold_store", "cold store loop, cold_head {}, hot_final_head {}", cold_head_height, hot_final_head_height);
+
+    if cold_head_height > hot_final_head_height {
+        tracing::error!(target: "cold_store", "Error, cold head is ahead of final head, this should never take place!");
```

Comment:
also a metric.

And shouldn't we return an error here then?

Reply:
also - print the heights of both values (will be useful in debugging).

tracing::error!(target: "cold_store", cold_head_height, hot_final_height, "Cold head is ahead - must never happen")